Fix exception handling in the prestub worker - take 2

janvorli opened the pull request:
There is a bug in the new exception handling when ThePreStub is called
from CallDescrWorkerInternal and the exception is propagated through
that. One of such cases is when a managed class is being initialized
during JITting and the constructor is in an assembly that's not found.
The bug results in skipping all the native frames upto a managed frame
that called that native chain that lead to the exception. In the
specific case I've mentioned, a lock in the native code is left in
locked state. That later leads to a hang. This was case was observed
with Roslyn invoking an analyzer where one of the dependencies was
missing.

The fix is to ensure that when ThePreStub is called by
CallDescrWorkerInternal, the exception is not caught in the
PreStubWorker. It is left flowing into the native calling code instead.

The same treatment is applied to ExternalMethodFixupWorker and
VSD_ResolveWorker too.

On Windows, we also need to prevent the ProcessCLRException invocation
to call into the managed exception handling code.

janvorli committed:
Fix exception handling in the prestub worker

There is a bug in the new exception handling when ThePreStub is called
from CallDescrWorkerInternal and the exception is propagated through
that. One of such cases is when a managed class is being initialized
during JITting and the constructor is in an assembly that's not found.
The bug results in skipping all the native frames upto a managed frame
that called that native chain that lead to the exception. In the
specific case I've mentioned, a lock in the native code is left in
locked state. That later leads to a hang. This was case was observed
with Roslyn invoking an analyzer where one of the dependencies was
missing.

The fix is to ensure that when ThePreStub is called by
CallDescrWorkerInternal, the exception is not caught in the
PreStubWorker. It is left flowing into the native calling code instead.

The same treatment is applied to ExternalMethodFixupWorker and
VSD_ResolveWorker too.

On Windows, we also need to prevent the ProcessCLRException invocation
to call into the managed exception handling code.

janvorli committed:
Fix missing ContextFlags setting

I was hitting intermittent crashes in the unhandled exception test
with GC stress C enabled due to this when GetSSP tried to access the
SSP in the extended part of the context.

janvorli commented on src/coreclr/vm/exceptionhandling.cpp:
```diff
@@ -932,14 +932,13 @@ ProcessCLRExceptionNew(IN     PEXCEPTION_RECORD   pExceptionRecord,
+    if (!ExecutionManager::IsManagedCode((PCODE)pDispatcherContext->ControlPc) ||
+        pThread->HasThreadStateNC(Thread::TSNC_ProcessedUnhandledException))
     {
-        if ((pExceptionRecord->ExceptionFlags & EXCEPTION_UNWINDING))
```
@jkotas this is the only change w.r.t. the previous PR that was reverted recently. This popping was causing the GC to get lost.

jkotas committed:
Merge branch 'main' into fix-exception-from-class-init-hang-3

jkotas commented:
The failure looks related:
```
Unhandled exception. System.OverflowException: Arithmetic operation resulted in an overflow.
   at Microsoft.Win32.OAVariantLib.OAFailed(Int32 hr)
   at Microsoft.Win32.OAVariantLib.ChangeType(Object source, Type targetClass, Int16 options, CultureInfo culture)
   at System.OleAutBinder.ChangeType(Object value, Type type, CultureInfo cultureInfo)
```

janvorli committed:
Fix couple of issues

* The previous set of changes removed popping of ExInfos too. That's not
  correct though. But it should be done in a different place than the
  ProcessCLRExceptionNew.
* There was a problem (even before the fix) that an exception caught in
  DispatchInfo::InvokeMember was reported (via console log and to the
  debugger) as unhandled.

This change also adds a new flavor of the unhandled exception test that
throws an unhandled exception on a secondary thread to exercise the
related code path.

janvorli committed:
Merge branch 'fix-exception-from-class-init-hang-3' of https://github.com/janvorli/runtime into fix-exception-from-class-init-hang-3

janvorli commented:
@jkotas I have fixed the test failure. There were actually two issues. One was that the exception was reported as unhandled even though it was actually then handled in native runtime code in the `DispatchInfo::InvokeMember`. The other was that the `ExInfo` popping was still needed and that lead to the test hang (in an interesting way - the next `ExInfo` that was pushed by the next portion of the test had accidentally the same address, so the presence of previous stale `ExInfo` link at the same address created an infinite loop in the `m_pPrevNestedInfo` chain.
